### ISSUE TITLE
[rllib] LSTM mit PPO action ouput and value function etc.

### ISSUE THREAD

**Issue by aGiant** [ray-project/ray#3799]
### System information
- **Ubuntu 18.04**:
- **Ray installed from pip3 install -U ray**:
- **Ray version: 0.6.2**:
- **Python 3.6**:

### Describe the problem

# Compute outputs of LSTM in LSTM.py
        last_layer = tf.reshape(lstm_out, [-1, cell_size])
        logits = linear(last_layer, num_outputs, "action",
                        normc_initializer(0.01))
        return logits, last_layer
I changed the linear() function in misc.py to:

```
def linear(x, size, name, initializer=None, bias_init=0):
    w = tf.get_variable(
        name + "/w", [x.get_shape()[1], size], initializer=initializer)
    b = tf.get_variable(
        name + "/b", [size], initializer=tf.constant_initializer(bias_init))
    return tf.sigmoid(tf.matmul(x, w) + b)
```

so that the action output stays in (0,1). But the outputs of actions were actually not in (0,1) and look like:
---------------------- 1997 ------------------------
[ 2.1607995   1.0698073  -1.3150089   1.5100739   0.49914685  1.1259065
  1.4430718  -0.5570196   3.2163894  -1.9132476  -0.9171299  -0.41201705
  0.71264625  1.2007536  -0.9179335   0.15855286 -0.9532902   1.3527832
 -1.3043945   2.9163842   1.699544  ]

Please help me here, many thanks!
@ericl

**Comment by ericl** [comment]
That seems like it should work, I would try using tf.print to debug. Note that you can also use a custom model.

Also be careful with squashing, it can easily cause gradients to disappear. Clipping seems more robust: https://github.com/ray-project/ray/pull/3426

**Comment by aGiant** [comment]
It should work but didn't, which means somehow LSTM didn't call the linear() but something else. If it was true then how could I ensure that the custom model is also correctly called in training. I used tf.print() to debug but successfully. Ray is very complicated and I don't have the whole picture of rllib structure. Please help me out of this problem, many many thanks!
